proposals/separate-get-set-field.md

```diff
@@ -0,0 +1,187 @@
+
+Obviously the above has absolutely no chance of supporting a reasonable
+equivalent `SetField`.
+
+## Proposed Change Specification
+
+`GHC.Records.HasField` would be adjusted from it's current planned
+specification:
+
+```haskell
+class HasField x r a | x r -> a where
+    hasField :: r -> (a -> r, a)
+```
+
+to become
+
+```haskell
+class GetField (x :: k) r a | x r -> a where
+    getField :: r -> a
+
+class GetField x r a => SetField (x :: k) r a | k r -> a where
+    setField :: a -> r -> r
+    setField = updateField @x . const
+
+    updateField :: (a -> a) -> r -> r
+    updateField f r = let (g, a) = hasField @x r in g (f a)
+
+    hasField :: r -> (a -> r, a)
+    hasField r = (\a -> setField @x a r, getField @x r)
+
+    {-# MINIMAL setField | updateField | hasField #-}
+```
+
+
+## Effect and Interactions
+
+This change would allow for the definition of various `GetField` instances
+that were mentioned above as desirable. This would then enable those instances
+to be utilized by dot-notation.
+
+This would not be backwards compatible with the existing implementation,
+but neither is the planned implementation. Unlike the planned implementation
+it would be a very mechanical rename to fix, as opposed to a potentially
+impossible definition of `hasField`.
+
+
+## Costs and Drawbacks
+
+The drawbacks seem minimal compared to the planned implementation. You can just
+treat `SetField` as the proposed `HasField` and everything will work the same.
+
+There is slightly more complexity due to having two separate classes rather
+than just one.
+
+
+## Alternatives
+
+Separate out further into "get-only", "set-only", "get-0-or-more",
+"get-and-set" etc. instead of only "get-only" and "get-and-set".
+
+The above makes a lot of sense for libraries like `lens` that deal with sums and
+clever multi-target lenses like `traverse`.
+
+However I do not believe it is particularly valuable for these `GHC.Records`
+classes, since they are focused on "fields" and "records" and automatically
+deriving them, and not on the general concept of lenses.
+
+Whenever you have a record that contains a `Maybe b` or a `[b]`, you will most
+likely want (and have derived for you) the single-target field from `a` to
+`Maybe b`/`[b]`, and not the multi/partial-target field from `a` to `b`. The
+latter you can always define yourself as a top level lens or as an instance of
+some other class.
+
+
+## Unresolved Questions
+
+Physical fields are currently resolved automatically by GHC whenever the field
+is in scope, so to allow for GHC to autoderive read-only fields we will want
+some syntax to export only the GetField instance and not the SetField instance.
+
+Should any of `hasField`, `setField`, `updateField` be dropped from `SetField`?
+They are all mentioned currently for performance reasons, as I can imagine
+situations where it seems like you need all 3 for optimal perf. But they can
+all be defined in terms of each other, so perhaps some should be dropped.
```

Comment:
I don't have a very strong opinion about which representation for setters we use (cunningly-disguised van Laarhoven lenses included). The situations in which there is a runtime performance difference are fairly arcane.

But I am slightly concerned that if the `SetField` class is too large, the compiler might need to do a lot of work generating the dictionaries, reducing compile-time performance. Bear in mind that we need to do this for all fields of all record types GHC compiles. @simonpj is that a reasonable concern?

Reply:
Implementation wise dictionaries could be created at use-site on the fly (current `HasField`), or at record definition site (cf. selector functions). 

Maybe even something one can control with some `-f` switch: if you go all in for overloaded fields machinery, at record definition site would be better. But for occasional use at use-site is probably fine.